File: python_cpp/uPIMulator_backend/src/simulator/cpu/cpu.cc
#include "simulator/cpu/cpu.h"

namespace upmem_sim::simulator::cpu {

CPU::~CPU() {
  delete init_thread_;
  delete sched_thread_;
  delete fini_thread_;
}

void CPU::connect_rank(rank::Rank* rank) {
  init_thread_->connect_rank(rank);
  sched_thread_->connect_rank(rank);
  fini_thread_->connect_rank(rank);
}

}  // namespace upmem_sim::simulator::cpu

File: python_cpp/uPIMulator_frontend/src/iss/cpu/cpu.py
from iss.cpu.fini_thread import FiniThread
from iss.cpu.init_thread import InitThread
from iss.cpu.sched_thread import SchedThread
from iss.dpu.dpu import DPU


class CPU:
    def __init__(self, benchmark: str, num_tasklets: int, dpu: DPU):
        self._init_thread: InitThread = InitThread(benchmark, num_tasklets, dpu)
        self._sched_thread: SchedThread = SchedThread(benchmark, num_tasklets, dpu)
        self._fini_thread: FiniThread = FiniThread(benchmark, num_tasklets, dpu)

    def num_executions(self) -> int:
        return self._sched_thread.num_executions()
        
    def init(self) -> None:
        self._init_thread.init()

    def launch(self) -> None:
        self._init_thread.launch()

    def sched(self, execution: int) -> None:
        self._sched_thread.sched(execution)

    def check(self, execution: int) -> None:
        self._sched_thread.check(execution)

    def fini(self) -> None:
        self._fini_thread.fini()

    def cycle(self) -> None:
        self._init_thread.cycle()
        self._sched_thread.cycle()
        self._fini_thread.cycle()
